system/settings/settings.c

```diff
@@ -0,0 +1,1717 @@
+ * Name: signotify
+ *
+ * Description:
+ *    Notify anything waiting for a signal
+ *
+ * Input Parameters:
+ *    none
+ *
+ * Returned Value:
+ *    none
+ *
+ ****************************************************************************/
+
+static void signotify(void)
+{
+  int i;
+
+  for (i = 0; i < CONFIG_SYSTEM_SETTINGS_MAX_SIGNALS; i++)
+    {
+      if (g_settings.notify[i].pid == 0)
+        {
+          break;
+        }
+
+      kill(g_settings.notify[i].pid, g_settings.notify[i].signo);
+    }
+}
+
+/****************************************************************************
+ * Name: dump_cache
+ *
+ * Description:
+ *    Writes out the cached data to the appropriate storage
+ *
+ * Input Parameters:
+ *    value            - parameter passed if called from a timer signal
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void dump_cache(union sigval ptr)
+{
+  int ret = OK;
+  FAR bool *wrpend = (bool *)ptr.sival_ptr;
+
+  int i;
+
+  ret = pthread_mutex_lock(&g_settings.mtx);
+  if (ret < 0)
+    {
+      DEBUGASSERT(0);
+    }
+
+  for (i = 0; i < CONFIG_SYSTEM_SETTINGS_MAX_STORAGES; i++)
+    {
+      if ((g_settings.store[i].file[0] != '\0') &&
+           g_settings.store[i].save_fn)
+        {
+          ret = g_settings.store[i].save_fn(g_settings.store[i].file);
+          if (ret < 0)
+            {
+              break;
+            }
+        }
+    }
+
+  *wrpend = false;
+
+  pthread_mutex_unlock(&g_settings.mtx);
+}
+
+/****************************************************************************
+ * Name: sanity_check
+ *
+ * Description:
+ *    Checks that the string does not contain "illegal" characters
+ *
+ * Input Parameters:
+ *    str              - the string to check
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+static int sanity_check(FAR char *str)
+{
+#ifdef CONFIG_DEBUG_ASSERTIONS
+  DEBUGASSERT(strchr(str, '=')  == NULL);
+  DEBUGASSERT(strchr(str, ';')  == NULL);
+  DEBUGASSERT(strchr(str, '\n') == NULL);
+  DEBUGASSERT(strchr(str, '\r') == NULL);
+#else
+  if ((strchr(str, '=')  != NULL) || (strchr(str, ';')  != NULL) ||
+      (strchr(str, '\n') != NULL) || (strchr(str, '\r') != NULL))
+    {
+      return -EINVAL;
+    }
+
+#endif
+
+  return OK;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: settings_init
+ *
+ * Description:
+ *   Initializes the settings storage.
+ *
+ * Input Parameters:  none
+ *
+ * Returned Value:    none
+ *
+ ****************************************************************************/
+
+int settings_init(void)
```

Comment:
I think at some point there was something that could return an error for some reason I now don't recall. But there is (now) nothing that fails so I will change it.

Reply:
It's not just about the return type.
I believe it will be good to retain the convention that init never fails.

Error handling before the system even begins can become stupidly complicated.